### PR TITLE
[Merged by Bors] - feat: change `IsFreeGroup` to a `Prop`

### PR DESCRIPTION
**Pull request by sgouezel**
Currently, the class `IsFreeGroup` contains data (namely, a specific set of generators). This is bad, as there are many sets of generators in a free group, and changing sets of generators happens all the time in geometric group theory. We switch to a design in which
* we define `FreeGroupBasis`, following the definition and API of bases of vector spaces. Most existing API around `IsFreeGroup` is transferred to lemmas taking a free group basis as a variable.
* The typeclass `IsFreeGroup` is Prop-valued, and requires only the existence of a free group basis.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/FreeGroup/IsFreeGroup.lean [hunk 1]
```diff
@@ -10,56 +10,177 @@ import Mathlib.GroupTheory.FreeGroup.Basic
...
+  nonempty_basis : ∃ (ι : Type u), Nonempty (FreeGroupBasis ι G)
+
+namespace FreeGroupBasis
+
+variable {ι ι' G H : Type*} [Group G] [Group H]
+
+/-- A free group basis for `G` over `ι` is associated to a map `ι → G` recording the images of
+the generators. -/
+instance funLike : FunLike (FreeGroupBasis ι G) ι (fun _ ↦ G) where
+  coe := fun b i ↦ b.repr.symm (FreeGroup.of i)
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/FreeGroup/IsFreeGroup.lean:72]: style nit:
```suggestion
  coe b := fun i ↦ b.repr.symm (FreeGroup.of i)
```
reads slightly more clearly as "the coercion of `b` is this function"

File: Mathlib/GroupTheory/FreeGroup/IsFreeGroup.lean [hunk 2]
```diff
@@ -10,56 +10,177 @@ import Mathlib.GroupTheory.FreeGroup.Basic
...
+  coe_injective' := by
+    rintro ⟨b⟩  ⟨b'⟩ hbb'
+    suffices H : (b.symm : FreeGroup ι →* G) = (b'.symm : FreeGroup ι →* G) by
+      have : b.symm = b'.symm := by ext x; exact FunLike.congr_fun H x
+      have : b.symm.symm = b'.symm.symm := by rw [this]
+      simpa using this
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/FreeGroup/IsFreeGroup.lean:78]: This should surely follow in one tactic from an existing injectivity lemma or a `symm_eq_symm` lemma or something?

**Review by eric-wieser — APPROVED** [review]
I haven't had time to review this in detail yet, but this approach looks great to me! I'd like to do something similar to `IsTensorProduct` at some point...

**Comment by riccardobrasca** [comment]
Just a comment about names: for module we have `Module.Free`. Should we use `isFree` everywhere?

File: Mathlib/GroupTheory/FreeGroup/IsFreeGroup.lean (hunk 2 above)
Comment (sgouezel) [reply Mathlib/GroupTheory/FreeGroup/IsFreeGroup.lean:78]: That's what I thought, but I couldn't locate a convenient lemma for that. Hence the slightly convoluted proof...

**Comment by sgouezel** [comment]
I don't really have an opinion about names. In this PR, I didn't want to change the name, so I kept `IsFreeGroup`, but I think it's a good name given that `FreeGroup` is already taken for the concrete construction.

**Review by Vierkantor — APPROVED** [review]
Apart from the missing `ext` lemma, this is a really neat refactor!

File: Mathlib/GroupTheory/FreeGroup/IsFreeGroup.lean (hunk 2 above)
Comment (eric-wieser) [reply Mathlib/GroupTheory/FreeGroup/IsFreeGroup.lean:78]: I think in general we have been avoiding adding the Equivalent versions of any ext lemmas; if you want them, you can get there by applying `MulEquiv.toMonoidHom_injective` or similar. Tagging *these* lemmas with `ext` may be reasonable.

**Pull request closed without merging**